VSR: Reset pipeline's prepare_ok quorums

sentientwaffle opened the pull request:
This fixes a correctness bug. When a view-change cycles around, so that the original leader becomes a leader of a new view, pipeline entries can be reused.

The pipeline's `prepare_ok` quorums must _not_ be reused — the replicas that sent them may have swapped them out during one of the previous view changes.

sentientwaffle committed:
VSR: Reset pipeline's prepare_ok quorums

This fixes a correctness bug. The pipeline's `prepare_ok` quorums must _not_ be reused between view cycles — the replicas that sent them may have swapped them out during one of the previous view changes.

When entries of the pipeline are reused between views, reset the `prepare_ok` quorum to avoid this issue:

1. Given a cluster of replicas `A,B,C,D,E`, with `A` as the initial leader:
2.`A` receives `prepare_ok` from `B` for op `5a`.
3. (Before any other `prepare_ok`s arrive, the view changes a couple times. `C` is now leader).
4. `C` rewrites op `5a` to `5b`. `B` replaces `5a` with `5b`.
5. (Before `5b` can be committed, the view changes a couple more times. `A` is leader again).
6. `A` chooses `5a` again, and it is already in the pipeline — but it (incorrectly) believes it has a valid `prepare_ok` from `B`.

`Replica.pipeline` is left untouched when a replica transitions from leader→follower (to improve availability via `on_request_prepare`). When a follower transitions back to leader, it will retain any entries of its pipeline that it would otherwise just read from disk during pipeline repair.

Those retained pipeline Prepares still have `prepare_ok` counters (`ok_from_all_replicas`) that may reflect outdated acks, because an op was rewritten and rewritten again (on replica `B` in this example).

jorangreef commented on src/vsr/replica.zig:
```diff
@@ -4550,7 +4550,7 @@ pub fn ReplicaType(
                 assert(replica < self.replica_count);
             }
 
-            counter.setIntersection(quorum_counter_null);
```
Just curious—was there a bug here, in using `setIntersection()`?

sentientwaffle replied:
No, there wasn't a bug there. I noticed that since these bitsets are values it is clearer to just assign `quorum_counter_null`.

jorangreef commented on src/vsr/replica.zig:
```diff
@@ -4569,6 +4569,17 @@ pub fn ReplicaType(
 
+        fn reset_quorum_prepare_ok(self: *Self) void {
+            // "prepare_ok"s from prior views are not valid, even if the pipeline entry is reused
+            // after a cycle of view changes.
```
```suggestion
            // "prepare_ok"s from previous views are not valid, even if the pipeline entry is reused
            // after a cycle of view changes. In other words, when a view change cycles around, so
            // that the original primary becomes a primary of a new view, pipeline entries may be
            // reused. However, the pipeline's prepare_ok quorums must not be reused, since the
            // replicas that sent them may have swapped them out during a previous view change.
```

jorangreef commented on src/vsr/replica.zig:
```diff
@@ -5276,6 +5289,8 @@ pub fn ReplicaType(
             var iterator = self.pipeline.iterator();
             while (iterator.next_ptr()) |prepare| {
                 assert(prepare.message.header.command == .prepare);
+                assert(!prepare.ok_quorum_received);
```
Just to double-check, are we vulnerable to an assertion crash here, for example, if a `prepare_ok` arrives in between resetting the quorum and before calling `verify_pipeline()`? Can this happen?

sentientwaffle replied:
No, that can't happen because:

- The replica resets the quorum when transitioning to `normal` or `view_change` status.
- `ignore_prepare_ok` rejects `prepare_ok`s when the replica is in `view_change`, so we only need to worry about `normal`.
- But `verify_pipeline` is called from:
  - `start_view_as_the_new_leader` (_before_ setting `status = normal`, so the `status` is still `view_change`).
  - And `repair_pipeline_diff`, which is only called during `status == view_change`.

So not only is it safe, but I can add `assert(self.status == .view_change)` to `verify_pipeline`.

jorangreef replied:
Nice!

jorangreef reviewed (approved):
Great find!

sentientwaffle committed:
VSR: Add assertion; clarify docs

In `verify_pipeline`, the `assert(!prepare.ok_quorum_received);` is safe because:

- The replica resets the quorum when transitioning to `normal` or `view_change` status.
- `ignore_prepare_ok` rejects `prepare_ok`s when the replica is in `view_change`, so we only need to worry about normal.
- But `verify_pipeline` is called from:
  - `start_view_as_the_new_leader` (before setting `status = normal`, so the `status` is still view_change).
  - And `repair_pipeline_diff`, which is only called during `status == view_change`.
